src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,351 @@
+        var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+        var result = await handler.ExecuteAsync(message, document, cancellationToken)
+            .ConfigureAwait(false);
+        var responseJson = JsonSerializer.Serialize(result, handler.ResponseType);
+        return responseJson;
+    }
+
+    public ValueTask UnregisterExtensionAsync(
+        string assemblyFilePath,
+        CancellationToken cancellationToken)
+    {
+        var assemblyFileName = Path.GetFileName(assemblyFilePath);
+        var assemblyFolderPath = Path.GetDirectoryName(assemblyFilePath)
+            ?? throw new InvalidOperationException($"Unable to get the directory name for {assemblyFilePath}.");
+
+        try
+        {
+            CustomMessageHandlerExtension? extension = null;
+            lock (_lockObject)
+            {
+                if (_extensions.TryGetValue(assemblyFolderPath, out extension))
+                {
+                    if (extension.Assemblies.TryGetValue(assemblyFileName, out var extensionAssembly))
+                    {
+                        extension.Assemblies.Remove(assemblyFileName);
+                        UnregisterHandlersForAssembly(extensionAssembly);
+                    }
+
+                    if (extension.Assemblies.Count == 0)
+                    {
+                        _extensions.Remove(assemblyFolderPath);
+                        foreach (var assembly in extension.Assemblies.Values)
+                        {
+                            UnregisterHandlersForAssembly(assembly);
+                        }
+                    }
+                }
+            }
+
+            extension?.AnalyzerAssemblyLoader.Dispose();
+        }
+        catch (Exception e) when (LogAndPropagate(e))
+        {
+            // unreachable
+        }
+
+        return ValueTask.CompletedTask;
+
+        bool LogAndPropagate(Exception e)
+        {
+            Logger.Log(
+                FunctionId.CustomMessageHandlerService_UnloadCustomMessageHandlerAsync,
+                $"Error unregistering {assemblyFilePath}: {e}",
+                LogLevel.Error);
+            return false;
+        }
+
+        void UnregisterHandlersForAssembly(CustomMessageHandlerAssembly? assembly)
+        {
+            if (assembly.HasValue)
+            {
+                foreach (var handler in assembly.Value.WorkspaceMessageHandlers)
+                {
+                    _handlers.Remove(handler);
+                }
+
+                foreach (var documentHandler in assembly.Value.DocumentMessageHandlers)
+                {
+                    _documentHandlers.Remove(documentHandler);
+                }
+            }
+        }
+    }
+
+    public ValueTask ResetAsync(CancellationToken cancellationToken)
+    {
+        Clear();
+        return ValueTask.CompletedTask;
+    }
+
+    public void Dispose()
+        => Clear();
```

Comment:
having this be both disposalbe and Clearable seems... odd.  why have dual semanitcs on supportin gthis.

Reply:
This must allow to be reset (it happens when a solution is closed) so that all loaded handlers are unloaded.
I figured that making the service disposable and unloading the loaded assemblies on dispose would be a good practice (this is why I have a shared behavior between `ResetAsync` and `Dispose`). I can make the service non-disposable, if you think it's not useful.

Reply:
i am 100% ok with Reset.  Seems necessary, given other code i saw around ILspLifecycle management.

DIsposing seems unnecessary.  We only dispose when literally teraing everything down.  So it would just be extra work.  Better to just let it all clean up normally.  It also simplifies things, which always makes me happy :)